qiskit/synthesis/boolean/boolean_expression_synth.py

```diff
@@ -0,0 +1,126 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2025.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Circuit synthesizers and related classes for boolean expressions"""
+
+import itertools
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.library import ZGate, XGate
+
+
+class EsopGenerator:
+    """Generates an ESOP (Exlusive-sum-of-products) representation
+    for a boolean function given by its truth table"""
+
+    def __init__(self, truth_table):
+        self.truth_table = truth_table
+        self.cofactor_table = {}
+        self.esop = self.generate_esop(assignment=())
+
+    def clause_diff(self, clause1, clause2):
+        """The indices of variables where the clauses differ"""
+        return [i for i in range(len(clause1)) if clause1[i] != clause2[i]]
+
+    def combine_and_optimize(self, clauses_1, clauses_2):
+        """Combining clauses of distance 1 until no more combinations can be performed"""
```

Comment:
One minor question: why do you pass `clauses_1` and `clauses_2` and not just `clauses_1 + clauses_2`?

Reply:
That's mostly a leftover from a previous version, in which I tried optimizing based on the fact that at the beginning, we are only interested in clauses obtained by `clause_1` from `clauses_1` and `clause_2` from `clauses_2`. It still makes sense to keep the interface as it is now, in my opinion; it gives more information than passing the combined clauses.